llvm/lib/SYCLLowerIR/ESIMD/LowerESIMDSlmReservation.cpp

```diff
@@ -0,0 +1,579 @@
+using SLMKernelUsageMap = DenseMap<
+    std::pair<const ScopedCallGraph::Node *, const ScopedCallGraph::FuncNode *>,
+    int>;
+using Node2IntMap = DenseMap<const ScopedCallGraph::ScopeNode *, int>;
+
+// Employs dynamic programming technique to find maximum SLM usage along all
+// paths in a scoped callgraph from given scope 'Cur' to given kernel 'Kernel'.
+// 'Kernel2MaxSLM' accumulates the maximum SLM usage per kernel (along any path
+// from the kernel to any scope), 'Results' is the dynamic programming result
+// cache, which later is also used to replace __esimd_slm_alloc calls with
+// constant SLM offset.
+int findMaxSLMUsageAlongAllPaths(const ScopedCallGraph::Node *Cur,
+                                 const ScopedCallGraph::FuncNode *Kernel,
+                                 Func2IntMap &Kernel2MaxSLM,
+                                 SLMKernelUsageMap &Results) {
+
+  // No protection from endless recursion in the algorithm, as recursion
+  // (cycles) in the call graph is prohibited.
+  auto ResI = Results.find({Cur, Kernel});
+
+  if (ResI != Results.end()) {
+    return ResI->second;
+  }
+  constexpr int KernelUnreachable = -1;
+  int MaxSLMUse = KernelUnreachable;
+  int SLMUseF = 0;
+
+  if (const auto *Scope = dyn_cast<ScopedCallGraph::ScopeNode>(Cur)) {
+    SLMUseF = getSLMUsage(Scope->getStart());
+  }
+  if (Cur == Kernel) {
+    MaxSLMUse = 0;
+  } else {
+    for (const auto &Pred : Cur->preds()) {
+      int SLMUse = findMaxSLMUsageAlongAllPaths(Pred.get(), Kernel,
+                                                Kernel2MaxSLM, Results);
+      MaxSLMUse = std::max(MaxSLMUse, SLMUse);
+    }
+  }
+  // If Kernel can not be reached via any of the predecessors, then discard
+  // SLMUseF as it does not affect total SLM size needed by the Kernel, and
+  // return -1.
+  int Res = MaxSLMUse < 0 ? MaxSLMUse : SLMUseF + MaxSLMUse;
+  Results[{Cur, Kernel}] = Res;
+  auto I = Kernel2MaxSLM.find(Kernel->getFunction());
+  assert(I != Kernel2MaxSLM.end());
+
+  // Update per-kernel maximum SLM usage.
+  if (I->second < Res) {
+    I->second = Res;
+  }
+  return Res;
+}
+
+size_t lowerSLMReservationCalls(Module &M) {
+  // Create a detailed "scoped" call graph. Scope start/end is marked with
+  // x = __esimd_slm_alloc / __esimd_slm_free(x)
+  ScopedCallGraph SCG(M);
+#ifndef NDEBUG
+  if (DebugLevel > 0) {
+    SCG.dump();
+  }
+#endif
+
+  // This maps a kernel to all reachable __esimd_slm_alloc calls. Each call is
+  // mapped to maximum prior SLM usage on any CG (reverse) path leading to the
+  // kernel.
+  SLMKernelUsageMap Results;
+  // Maps a kernel to the maximum SLM size it can potentially use.
+  Func2IntMap Kernel2MaxSLM;
+  // Maps a scope node to maximum prior SLM usage on any (reverse) path leading
+  // to any kernel.
+  Node2IntMap Scope2MaxSLM;
+
+  // Now, for each <ScopeNode, kernel FuncNode> pair:
+  // find all possible (reverse) paths in the graph in the graph from the
```

Comment:
"in the graph in the graph":  2x "in the graph".

Reply:
ok